### PR TITLE
feat(plugin): Support service-specific options

### PR DESCRIPTION
**Pull request by NickyMeuleman**
**What**:
This is a feature request in the form of a PR to add support for custom URL parameters when fetching the HTML from twitter.
If preferred, I think I can make a codeSandbox with these changes and make an issue referring to that.
references #25

A list of supported parameters can be found here:
https://developer.twitter.com/en/docs/twitter-for-websites/embedded-tweets/guides/embedded-tweet-parameter-reference
for the specific URL used in this project:
https://developer.twitter.com/en/docs/tweets/post-and-engage/api-reference/get-statuses-oembed

**Why**: The HTML was already being fetched by sending a `fetch` request that included some parameters. This allows the user to expand/overwrite those in their `gatsby-config`
More options = more better, right? 🤡 

**How**: Passing the plugin options (provided in `gatsby-config`) down to the `getHTML` function.
The HOW can change, this PR is meant as a proof-of-concept.

possible `gatsby-config`
```js
plugins: [
  {
    resolve: `gatsby-transformer-remark`,
    options: {
      plugins: [
        {
          resolve: `gatsby-remark-embedder`,
          options: {
            twitter: {
				"theme": "dark",
				"hide_thread": true
			},
          },
        },
      ],
    },
  },
];
```

Edit: Same principle could be used to send additional url parameters to the Instagram oembed url (after the PR adding Instagram is merged)
https://www.instagram.com/developer/embedding/

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

**Comment by MichaelDeBoey** [comment]
Hi @NickyMeuleman!

Are you still interested in implementing this POC?
Would be awesome to see some progress on this one. 🙂

**Comment by NickyMeuleman** [comment]
Absolutely! Are you ok with the basic implementation of this (an object per provider with possible url parameters, that gets passed down to the method responsible for getting the correct HTML)?

**Comment by MichaelDeBoey** [comment]
@NickyMeuleman I'm not quite sure how I would want to have the API at this moment and what parameters we would want to support tbh.
Also not sure if we need to pass the _whole_ config or just the config that matters for that exact transformer.

Maybe @kentcdodds has some valuable insights in what he thinks would be a good first approach? 🤔

**Comment by NickyMeuleman** [comment]
@MichaelDeBoey can you follow me on Twitter so I can DM you and talk about some ideas for how to make this work for every transformer and not just Twitter?
(ofc any useful info will make its way into this PR)
https://twitter.com/NMeuleman
(Also, I'm excited about seeing someone else that speaks Dutch that is into Gatsby :D)

**Comment by NickyMeuleman** [comment]
Rebased so the Instagram PR is included and url-parameters can be supplied for Instagram as well.

Added some docs.
Tested if the oEmbed urls are being built correctly.
Deliberately did not test if/how those urls worked (that responsability lies with the oEmbed provider)

**Comment by kentcdodds** [comment]
I have an idea. Hold on

**Comment by MichaelDeBoey** [comment]
Thanks for all the work you've been doing on this PR @NickyMeuleman! 👊

I'm still not sure what direction I want to go out with this tbh 🤔

The main questions I keep doubting about are
- How to pass the config object to the transformers?
  - Should we pass the whole config object (maybe without the `customTransformers` key)?
  - Should we only pass the part that's specific for that transformer?
    This would mean that a transformer can't use values of other transformers to calculate it's own config.  
    I can't directly think of a good example other then one that combines 2 other transformers tho
- What config values do we want to support?
  - Only a couple of parameters that are the same for each transformer (i.e. `width`, `height`, ...)?
  - Also custom parameter that are specific for the transformer (i.e. `theme` for CodeSandbox, ...)?

Let's wait and see what @kentcdodds is going to suggest while we can have some thought about my questions above.

**Comment by kentcdodds** [comment]
I think this is fine, except I would probably change it to be nested under a named option. That way we can add more options in the future if we need. And it's more clear.

```js
plugins: [
  {
    resolve: `gatsby-transformer-remark`,
    options: {
      plugins: [
        {
          resolve: `gatsby-remark-embedder`,
          options: {
            twitter: {
                       queryParams: {
				"theme": "dark",
				"hide_thread": true
                       }
			},
          },
        },
      ],
    },
  },
];
```

**Comment by NickyMeuleman** [comment]
Done!
Also slightly reworked how the documentation for this is done.
Now, the options part of each service can house a clear table listing which options are supported, not specific to url-parameters (maybe my next PR is adding those for codepen).

**Comment by MichaelDeBoey** [comment]
I'm not  quite sure about the `queryParams` name tbh.

This implies we're always doing a request to some endpoint,  but most cases we're just doing a manual transform and we want to be able to set some values (i.e. `height` & `width`).

I can also imagine that instead of setting `height` & `width`, we just want to set `ratio` for  a YouTube video for instance.
That would mean we have some manual work to do that has nothing to do with the query of the request.

**Comment by NickyMeuleman** [comment]
Support for options in other services doesn't need to use it.

An example of how a CodePen config might look
```js
plugins: [
  {
    resolve: `gatsby-transformer-remark`,
    options: {
      plugins: [
        {
          resolve: `gatsby-remark-embedder`,
          options: {
            codepen: {
				height: 300,
				border: "thin"
			},
          },
        },
      ],
    },
  },
];
```
official docs: https://blog.codepen.io/documentation/features/embedded-pens/

The options that are possible are listed in the documentation per service.
Only services that use query parameters will have the `queryParams` key.

**Comment by MichaelDeBoey** [comment]
I'm not really fond of that approach.

The way I see it is that the request or the manual composition is an implementation detail.
If we're going to do it in 2 different ways (i.e. normal keys vs `queryParams`), we'll expose this.
This will also cause a breaking change whenever we decide to use a request to get the html for CodePen for instance, otherwise it won't be consistent anymore with the other parameters.

**Comment by kentcdodds** [comment]
Sorry @MichaelDeBoey, I hadn't seen your comment until just now.

I'm thinking maybe it would be better to [invert control](https://kcd.im/ioc) here and allow users to override the built-in transforms somehow. Or maybe that's already possible?

**Comment by MichaelDeBoey** [comment]
@kentcdodds It's not possible to completely override the built-in transformers at this moment.

---

1. I'm all in favor of making it possible to add parameters to the transformer, so the user can change `height`, `theme`, `width`, ...

2. I don't like the fact that for `Instagram`, `Streamable` & `Twitter` we're using the `queryParams` object
```js
{
  resolve: 'gatsby-remark-embedder',
  options: {
    twitter: {
      queryParams: {
        hide_thread: true,
        theme: 'dark',
      },
    },
  },
},
```
And  for all other  transformers, we're putting these options directly under the transformer namespace
```js
{
  resolve: 'gatsby-remark-embedder',
  options: {
    codepen: {
      height: 300,
      border: 'thin',
    },
  },
},
```

3. I also like the idea of having a separate object for these values.

That's why I think we should put all options/params in an object, but we should think about a better name than `queryParams` so it has a meaningful name that isn't tied to the fact that we're passing the object as query params (see `Twitter`, ...) or we manually set these in the HTML (see all other transformers).
This will also have the benefit that we can swap between a `fetch` call and doing it manually without having a breaking change.

---

Maybe a bit out of scope for this PR, but since I think this can cause a breaking change, I think we should take this also into account:

Something else I'm still not sure about is the fact that I want to introduce a `ratio` option  for all video transformers.
So that we can make them responsive without setting `height` &  `width` manually.

So that raises the following questions:
- Should we just pass all the query parameters the user sets?
- Or should we check if `ratio` is  set and act on that?

**Comment by kentcdodds** [comment]
That's well-reasoned. I trust you'll make a good choice 👍

**Comment by NickyMeuleman** [comment]
What actions should I take to move this PR forward?

**Comment by MichaelDeBoey** [comment]
@NickyMeuleman I think we should _always_ put the options/parameters in an object whether or not they're passed to the `fetchOEmbedData` call or not.

Just have to figure out a meaningful name that works for both cases.

Also rebase your branch onto `master/upstream` to resolve conflicts 🙂

**Comment by osiux** [comment]
Why not just `òptions` or `params`? I think both of them work for both cases.

**Comment by MichaelDeBoey** [comment]
I'm more in favor of `params` I think, since `options` is to broad and could include the `ratio` too.

**Comment by NickyMeuleman** [comment]
This rebase was an ... _adventure_ but in the end everything is there.
Removed the `queryParams`nqme again as discussed above.
I believe this PR is ready to be merged now and discussions for future options can continue (should I open an issue for that?)

**Comment by NickyMeuleman** [comment]
Rebased again.
Should the options be wrapped in a `params` object @MichaelDeBoey ?

Understood
> I think we should always put the options/parameters in an object whether or not they're passed to the fetchOEmbedData call or not.
as "each service has their own key in the options".

I guess why this is still here is because you meant "each service has their own key and inside that key there is another `params` key if applicable"?

**Comment by NickyMeuleman** [comment]
Nested the url parameters inside a `params` key and documented usage.

Getting `Identifier 'hide_media' is not in camel case  babel/camelcase` when running `npm run validate` whihch is why the check on github is failing.

Any advice on how to deal with those?

File: src/transformers/Twitter.js [hunk 1]
```diff
@@ -1,23 +1,35 @@
...
-    `https://publish.twitter.com/oembed?url=${url}&dnt=true&omit_script=true`
-  ).then(({ html }) =>
+export const buildUrl = (url, options = {}) => {
+  // https://developer.twitter.com/en/docs/tweets/post-and-engage/api-reference/get-statuses-oembed
+  const urlObj = new URL(`https://publish.twitter.com/oembed`);
+  urlObj.search = new URLSearchParams({
+    url,
+    dnt: true,
+    omit_script: true,
+    ...options.params,
```
Comment (NickyMeuleman) [inline src/transformers/Twitter.js]: the babel/no-camelcase errors are coming from here (same with Instagram) if you passs an option that is not camelcase (like "hide_media").

How should this be handled?

**Comment by MichaelDeBoey** [comment]
@NickyMeuleman After giving this some thoughts, I think we should have an `services` key inside the plugin options (that lives next to the `customTransformers` key) and within that object having a key for each transformer like:

```js
{
  resolve: 'gatsby-remark-embedder',
  options: {
    customTransformers: [
      // ...
    ],
    services: {
      codepen: {
        // ...
      },
      // ...
    },
  },
},
```

This way we don't pollute the global plugin options with each new service we support.

I now also think we should just keep all service settings into the root of that service's settings object, so we can have something like:
```js
egghead: {
  affiliate: '5236ad',
  preload: true,
},
pinterest: {
  board: {
    pinSize: 'square',
  },
  pin: {
    pinSize: 'small',
    terse: false,
  },
  profile: {
    pinSize: 'square',
  },
},
twitter: {
  theme: 'light',
},
```

As last thing, I would just implement support for service-specific options and create a new PR for each service separately so we can discuss all options over there.

I'm sorry it's taking so long, but I really want to get this one properly implemented.

**Comment by NickyMeuleman** [comment]
Implemented mechanism to pass plugin-specific options to the transformers.
Added a todo test named `'passes service-specific options to the transformers'`

**Comment by NickyMeuleman** [comment]
Is this okay?

File: src/__tests__/plugin.js [hunk 2]
```diff
@@ -69,4 +71,40 @@ describe('gatsby-remark-embedder', () => {
...
+        shouldTransform: () => true,
+      };
+
+      const markdownAST = getMarkdownASTForFile('ServiceTransformer', true);
+
+      await plugin(
+        { cache, markdownAST },
+        {
+          customTransformers: [transformer],
+          services: { [name]: passedOptions },
```
Comment (MichaelDeBoey) [inline src/__tests__/plugin.js]:
```suggestion
          services: { serviceTransformer: { service: 'transformer' } },
```

File: src/__tests__/plugin.js [hunk 3]
```diff
@@ -69,4 +71,40 @@ describe('gatsby-remark-embedder', () => {
...
+      expect(transformer.getHTML).toHaveBeenCalledWith(
+        'https://some-site.com/id/abc',
+        passedOptions
+      );
+    },
+    {
+      'transformer with name and options': {
+        name: 'ServiceTransformer',
+        passedOptions: { like: 'turtles' },
+      },
```
Comment (MichaelDeBoey) [inline src/__tests__/plugin.js]:
```suggestion
      'transformer with name': {
        name: 'serviceTransformer',
        passedOptions: { service: 'transformer' },
      },
```

File: src/__tests__/plugin.js [hunk 4]
```diff
@@ -69,4 +71,40 @@ describe('gatsby-remark-embedder', () => {
...
+      'transformer without options': {
+        name: 'SomeTransformer',
+      },
```
Comment (MichaelDeBoey) [inline src/__tests__/plugin.js]:
```suggestion
      'transformer without name': {
        passedOptions: {},
      },
```

File: src/__tests__/plugin.js [hunk 5]
```diff
@@ -69,4 +71,40 @@ describe('gatsby-remark-embedder', () => {
...
+      'transformer without name or options': {},
```
Comment (MichaelDeBoey) [inline src/__tests__/plugin.js]:
```suggestion
```

File: src/index.js [hunk 6]
```diff
@@ -46,12 +46,12 @@ export default async (
 
     transformers
       .filter(({ shouldTransform }) => shouldTransform(urlString))
-      .forEach(({ getHTML }) => {
+      .forEach(({ getHTML, name }) => {
```
Comment (MichaelDeBoey) [inline src/index.js]:
```suggestion
      .forEach(({ getHTML, name = '' }) => {
```

File: src/index.js [hunk 7]
```diff
@@ -46,12 +46,12 @@ export default async (
...
         transformations.push(async () => {
           let html = await cache.get(urlString);
 
           if (!html) {
-            html = await getHTML(urlString);
+            html = await getHTML(urlString, services[name]);
```
Comment (MichaelDeBoey) [inline src/index.js]:
```suggestion
            html = await getHTML(urlString, services[name] || {});
```

**Comment by MichaelDeBoey** [comment]
@all-contributors Please add @NickyMeuleman for code

**Pull request merged**